File: lib/auth/register_user_use_case.dart
import 'package:diplomovka/auth/auth_service.dart';
import 'package:firebase_auth/firebase_auth.dart';

// Use Case pre registráciu používateľa pomocou e-mailu a hesla.
class RegisterUserUseCase {
  final AuthService _authService;

  RegisterUserUseCase(this._authService);

  /// Vykoná registráciu používateľa a vráti objekt User.
  Future<User?> execute(String email, String password) async {
    try {
      // Zavoláme metódu z Data Layer na vykonanie registrácie
      // (ktorá už obsahuje aj odoslanie overovacieho e-mailu)
      final user = await _authService.registerWithEmailAndPassword(email, password);
      return user;

    } on FirebaseAuthException {
      // Necháme chybu prejsť ďalej do ViewModelu, kde ju spracujeme.
      rethrow;
    }
  }
}

File: app/src/main/java/com/example/diplomovka_kotlin/domain/RegisterUserUseCase.kt
package com.example.diplomovka_kotlin.domain

import com.example.diplomovka_kotlin.data.services.AuthService
import com.google.firebase.auth.FirebaseAuthException
import com.google.firebase.auth.FirebaseUser

class RegisterUserUseCase(private val authService: AuthService) {

    suspend fun execute(email: String, password: String): FirebaseUser? {
        return try {
            authService.registerWithEmail(email, password)
        } catch (e: FirebaseAuthException) {
            throw e
        }
    }
}
